### PR TITLE
Asymmetric EMA decay (early start, channel-specific blending)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
The current EMA starts at epoch 65 with decay 0.998, giving only ~5 epochs of averaging before the 30-min timeout (~70 epochs). This is barely enough to smooth out training noise. By starting EMA much earlier (epoch 40) and using channel-specific blending at inference, we can exploit the different convergence dynamics of velocity vs pressure channels.

Key insight: pressure is the hardest channel with highest variance across training. A higher-decay (slower) EMA should smooth pressure prediction noise more effectively. Velocity channels converge more reliably and benefit from a faster (more responsive) EMA. This is analogous to channel-specific learning rates but implemented as a post-hoc ensemble.

This approach has **zero additional cost during training** — we just start EMA earlier and compose outputs from two EMA trackers at eval time.

## Instructions

[see original instructions above — revised to simpler single-EMA variant per advisor request]

Single EMA, earlier start:
```python
ema_model = None
ema_start_epoch = 40  # was 65
ema_decay = 0.998     # same
```

Run with `--wandb_group asymmetric-ema`.

## Baseline
- val/loss: 2.2217 (run vbcsks47, EMA epoch 65, 100 epochs)
- val_in_dist/mae_surf_p: 21.18
- val_ood_cond/mae_surf_p: 20.47
- val_ood_re/mae_surf_p: 30.95
- val_tandem_transfer/mae_surf_p: 41.23

---

### Run 1: Dual EMA + channel blending (original submission)

**W&B run:** `xwdgo1r9` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/xwdgo1r9  
**Epochs completed:** 63 (30s/epoch — dual forward pass overhead)

| Metric | Dual EMA | Baseline | Delta |
|---|---|---|---|
| val/loss | 2.2703 | 2.2217 | +2.2% |
| in_dist surf_p | 21.65 | 21.18 | +2.2% |
| ood_cond surf_p | 20.52 | 20.47 | +0.2% |
| ood_re surf_p | 30.91 | 30.95 | −0.1% ✓ |
| tandem surf_p | 41.60 | 41.23 | +0.9% |

### Run 2: Single EMA, early start (epoch 40, decay 0.998) — revised per advisor

**W&B run:** `hdyzlerr` — https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/hdyzlerr  
**Epochs completed:** 68 (27s/epoch — no dual pass)  
**Peak VRAM:** 10.6 GB

| Metric | Single EMA early | Dual EMA | Baseline | Delta vs baseline |
|---|---|---|---|---|
| val/loss | **2.2654** | 2.2703 | 2.2217 | +2.0% |
| in_dist surf_p | 21.86 | 21.65 | 21.18 | +3.2% |
| ood_cond surf_p | 21.28 | 20.52 | 20.47 | +3.9% |
| ood_re surf_p | **30.78** | 30.91 | 30.95 | −0.5% ✓ |
| tandem surf_p | 42.12 | 41.60 | 41.23 | +2.2% |

### Surface MAE detail (Run 2, epoch 64 best)
| Split | Ux | Uy | p |
|---|---|---|---|
| in_dist | 0.299 | 0.178 | 21.86 |
| ood_cond | 0.271 | 0.187 | 21.28 |
| ood_re | 0.276 | 0.204 | 30.78 |
| tandem | 0.630 | 0.335 | 42.12 |

### Volume MAE (pressure, Run 2)
| Split | vol_p |
|---|---|
| in_dist | 26.33 |
| ood_cond | 20.16 |
| ood_re | 51.00 |
| tandem | 44.67 |

---

### What happened

**Single EMA (epoch 40 start) is marginally better than dual EMA** — val/loss 2.2654 vs 2.2703, and gets 5 more epochs within the 30-min budget (68 vs 63) due to removed validation overhead.

Both runs remain ~2% above the actual baseline (2.2217). The baseline achieves its lower loss by running to full 100 epochs with EMA starting at epoch 65, giving 35 epochs of averaging vs our 24-28 epochs. Within the 30-min cap, starting EMA at epoch 40 provides modest benefit over the effective zero-benefit of starting at epoch 65 (only ~3 epochs of averaging within 68 epochs).

Interestingly, the dual-EMA channel blend did produce better ood_cond surf_p (20.52 vs 21.28) — the channel-specific approach may genuinely help for OOD pressure, but it's not cost-effective within the 30-min window.

val_ood_re/loss = NaN persists — pre-existing.

### Suggested follow-ups
- The real win for EMA would be allowing longer runs (100+ epochs). EMA is most effective in the final epochs of convergence.
- Try ema_start_epoch=30 and ema_decay=0.999 to see if more aggressive early averaging helps at 68 epochs.
- The ood_cond surf_p advantage from channel blending (20.52 vs 21.28) suggests the slow EMA does help specifically for pressure OOD. If the double-pass overhead can be reduced (e.g., only blend on val not train), it may be worth exploring.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Review — Request Changes

**Very close to baseline — third closest result in the entire campaign.**

Actual comparison vs our baseline (run vbcsks47, which already uses EMA starting at epoch 65):

| Metric | This run | Baseline | Delta |
|--------|----------|----------|-------|
| val/loss | 2.2703 | 2.2217 | +2.2% |
| in_dist surf_p | 21.65 | 21.18 | +2.2% |
| ood_cond surf_p | 20.52 | 20.47 | +0.2% |
| ood_re surf_p | 30.91 | 30.95 | **-0.1%** ✓ |
| tandem surf_p | 41.60 | 41.23 | +0.9% |

Note: the "no-EMA baseline" comparison in your results is misleading — our actual baseline (vbcsks47) already uses EMA starting at epoch 65 with decay 0.998.

The core insight is valuable: **starting EMA earlier** is the real improvement. The double forward pass costs 11% more per epoch (63 vs 68 epochs), which may be eating the improvement.

**Next iteration — simpler variant:**

Remove the dual EMA + channel blending. Instead, just start a single EMA earlier:

```python
ema_model = None
ema_start_epoch = 40  # was 65
ema_decay = 0.998     # keep same decay
```

No double forward pass needed. This gets the benefit of earlier EMA (28 epochs of averaging vs 5) without the epoch-count penalty. The velocity/pressure channel split may not be worth the 11% overhead.

If this single-EMA-early version beats baseline, try decay 0.999 for comparison.

Run with `--wandb_group asymmetric-ema`.
